llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -1722,45 +1730,118 @@ Value *LibCallSimplifier::optimizeNew(CallInst *CI, IRBuilderBase &B,
   uint8_t HotCold;
   if (CI->getAttributes().getFnAttr("memprof").getValueAsString() == "cold")
     HotCold = ColdNewHintValue;
+  else if (CI->getAttributes().getFnAttr("memprof").getValueAsString() ==
+           "notcold")
+    HotCold = NotColdNewHintValue;
   else if (CI->getAttributes().getFnAttr("memprof").getValueAsString() == "hot")
     HotCold = HotNewHintValue;
   else
     return nullptr;
 
+  // For calls that already pass a hot/cold hint, only update the hint if
+  // directed by OptimizeExistingHotColdNew. For other calls to new, add a hint
+  // if cold or hot, and leave as-is for default handling if "notcold" aka warm.
```

Comment:
If an existing hinted allocation is to be marked with a notcold hint, there should be a slight performance advantage to delegate to an unhinted variant rather than passing in the new hinted value. It's not a very important use case right now  but perhaps worth noting in the comment?

Reply:
> there should be a slight performance advantage to delegate to an unhinted variant rather than passing in the new hinted value

I might be missing something, but why would that be? Afaik tcmalloc should treat something with the default notcold hint value the same as anything else that isn't cold hinted. Or are you just talking about the need for the allocator to check the hint value being passed in?

Reply:
Yes, I'm referring to the latter. There is an additional load (of a parameter value), a check and a branch which can be avoided if we replace it with the unhinted variant.

Reply:
Added comment